services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java

```diff
@@ -18,62 +18,199 @@
 import com.futurewei.alcor.common.db.CacheException;
 import com.futurewei.alcor.common.db.CacheFactory;
 import com.futurewei.alcor.common.db.ICache;
-import com.futurewei.alcor.common.db.repo.ICacheRepository;
+import com.futurewei.alcor.common.db.Transaction;
+import com.futurewei.alcor.portmanager.entity.PortNeighbors;
+import com.futurewei.alcor.web.entity.dataplane.NeighborInfo;
 import com.futurewei.alcor.web.entity.port.PortEntity;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.context.annotation.ComponentScan;
 import org.springframework.stereotype.Repository;
 import javax.annotation.PostConstruct;
+import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
 @ComponentScan(value="com.futurewei.alcor.common.db")
 @Repository
-public class PortRepository implements ICacheRepository<PortEntity> {
+public class PortRepository {
     private static final Logger LOG = LoggerFactory.getLogger(PortRepository.class);
 
-    private ICache<String, PortEntity> cache;
+    private ICache<String, PortEntity> portCache;
+    private ICache<String, PortNeighbors> neighborCache;
 
     @Autowired
     public PortRepository(CacheFactory cacheFactory) {
-        cache = cacheFactory.getCache(PortEntity.class);
+        portCache = cacheFactory.getCache(PortEntity.class);
     }
 
     @PostConstruct
     private void init() {
         LOG.info("PortRepository init done");
     }
 
-    @Override
-    public PortEntity findItem(String id) throws CacheException {
-        return cache.get(id);
+    public PortEntity findPortEntity(String portId) throws CacheException {
+        return portCache.get(portId);
     }
 
-    @Override
-    public Map<String, PortEntity> findAllItems() throws CacheException {
-        return cache.getAll();
+    public Map<String, PortEntity> findAllPortEntities() throws CacheException {
+        return portCache.getAll();
     }
 
-    @Override
-    public void addItem(PortEntity portEntity) throws CacheException {
-        cache.put(portEntity.getId(), portEntity);
+    public void createPortAndNeighbor(PortEntity portEntity, NeighborInfo neighborInfo) throws Exception {
+        try (Transaction tx = portCache.getTransaction().start()) {
```

Comment:
Lock on portCache may not be sufficient. If multiple ports (in the same vpc) are created in separate calls, those calls will operate on different portCache row but the same neighborCache row.

Reply:
This lock can lock multiple cache, It should be taken out of the cache.

Reply:
This lock can lock multiple caches, It should be taken out of the cache.

Reply:
oh could you pls point me to the right codes that explain a bit more about multi-cache locking? Just out of curiosity.

Reply:
https://github.com/futurewei-cloud/alcor/blob/master/lib/src/main/java/com/futurewei/alcor/common/db/ignite/IgniteTransaction.java